chainer/functions/pooling/roi_pooling_2d.py

```diff
@@ -0,0 +1,334 @@
+# Modified work:
+# -----------------------------------------------------------------------------
+# Copyright (c) 2015 Preferred Infrastructure, Inc.
+# Copyright (c) 2015 Preferred Networks, Inc.
+# -----------------------------------------------------------------------------
+
+# Original work of _roi_pooling_slice, forward_cpu and backward_cpu:
+# -----------------------------------------------------------------------------
+# Copyright 2014 Nervana Systems Inc.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# -----------------------------------------------------------------------------
+
+# Original work of forward_gpu and backward_gpu:
+# -----------------------------------------------------------------------------
+# Fast R-CNN
+# Copyright (c) 2015 Microsoft
+# Licensed under The MIT License [see fast-rcnn/LICENSE for details]
+# Written by Ross Girshick
+# -----------------------------------------------------------------------------
+
+import numpy
+import six
+
+from chainer import cuda
+from chainer import function
+from chainer.utils import type_check
+
+
+def _roi_pooling_slice(size, stride, max_size, roi_offset):
+    start = int(numpy.floor(size * stride))
+    end = int(numpy.ceil((size + 1) * stride))
+
+    start = min(max(start + roi_offset, 0), max_size)
+    end = min(max(end + roi_offset, 0), max_size)
+
+    return slice(start, end), end - start
+
+
+class ROIPooling2D(function.Function):
+
+    """RoI pooling over a set of 2d planes."""
+
+    def __init__(self, outh, outw, spatial_scale):
+        self.outh, self.outw = outh, outw
+        self.spatial_scale = spatial_scale
+
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == 2)
+
+        x_type, roi_type = in_types
+        type_check.expect(
+            x_type.dtype == numpy.float32,
+            x_type.ndim == 4,
+            roi_type.dtype == numpy.float32,
+            roi_type.ndim == 2,
+            roi_type.shape[1] == 5,
+        )
+
+    def forward_cpu(self, inputs):
+        bottom_data, bottom_rois = inputs
+        n_rois, channels, height, width = bottom_data.shape
+        top_data = numpy.empty((n_rois, channels, self.outh, self.outw),
+                               dtype=numpy.float32)
+        self.argmax_data = numpy.empty_like(top_data).astype(numpy.int32)
+
+        for i_roi in six.moves.range(n_rois):
+            idx, xmin, ymin, xmax, ymax = bottom_rois[i_roi]
```

Comment:
From this line, we must impose `x_type.shape[0] == roi_type.shape[0]` in `check_type_forward`